### PR TITLE
Fixes 031215

### PR DESCRIPTION
**Pull request by ikeough**
This PR represents adjustments to Dynamo core required for [this Revit PR](https://github.com/DynamoDS/DynamoRevit/pull/242)

PTAL:
@pboyer

### CODE REVIEW

File: src/Libraries/CoreNodeModels/DropDown.cs [hunk 1]
```diff
@@ -104,7 +105,12 @@ protected override void DeserializeCore(XmlElement nodeElement, SaveContext cont
...
                 return;
 
-            SelectedIndex = ParseSelectedIndex(attrib.Value, Items);
+            selectedIndex = ParseSelectedIndex(attrib.Value, Items);
+
+            if (selectedIndex < 0)
+            {
+                Warning(Resources.NothingIsSelectedWarning);
+            }
         }
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/DropDown.cs]: What does this have to do with this PR?

File: src/DynamoCore/Interfaces/IStartConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        string Context { get; set; }
+        string DynamoCorePath { get; set; }
+        IPreferences Preferences { get; set; }
+        bool StartInTestMode { get; set; }
+        IUpdateManager UpdateManager { get; set; }
+        ISchedulerThread SchedulerThread { get; set; }
+        string GeometryFactoryPath { get; set; }
+        IAuthProvider AuthProvider { get; set; }
+        string PackageManagerAddress { get; set; }
+    }
```
Comment (pboyer) [inline src/DynamoCore/Interfaces/IStartConfiguration.cs]: See DynamoRevit for a recommended pattern here.  You might also fix the DynamoViewModel so that they are consistent.

File: src/Libraries/CoreNodeModels/DropDown.cs (hunk 1 above)
Comment (ikeough) [inline src/Libraries/CoreNodeModels/DropDown.cs]: This is fixes for tests from today as well as the refactoring of the StartConfigurations.

File: src/DynamoCore/Interfaces/IStartConfiguration.cs (hunk 2 above)
Comment (ikeough) [inline src/DynamoCore/Interfaces/IStartConfiguration.cs]: You are correct. I fixed this in the Revit PR, but I hadn't edited it here after we talked. It's up to date now.

File: test/Libraries/TestServices/TestSessionConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    /// reading the TestServices configuration file, if provided,
+    /// to provide data for configuring a test session from a location
+    /// other than Dynamo's core directory.
+    /// </summary>
+    public class TestSessionConfiguration
+    {
+        private const string CONFIG_FILE_NAME = "TestServices.dll.config";
+
+        public string DynamoCorePath { get; set; }
+        public LibraryVersion RequestedLibraryVersion { get; set; }
```
Comment (Benglin) [inline test/Libraries/TestServices/TestSessionConfiguration.cs]: These are no longer `private set`, their values are assigned during construction, any caller that requires these to be set should be redesigned, IMO. If we allow those callers to alter these properties, then the process is no longer centralized, those callers will have their own logic and once `TestSessionConfiguration` needs updating, those other parts better be updated.

Comment (ikeough) [inline test/Libraries/TestServices/TestSessionConfiguration.cs]: You are correct Ben. Even Resharper was begging me to privatize those. Done.

File: test/Libraries/TestServices/AssemblyResolver.cs [hunk 4]
```diff
@@ -21,8 +21,8 @@ public static void Setup()
         {
             if (resolverSetup) return;
 
-            var remoteConfig = new RemoteTestSessionConfig();
-            DynamoPathManager.Instance.InitializeCore(remoteConfig.DynamoCorePath);
+            var testConfig = new TestSessionConfiguration();
+            DynamoPathManager.Instance.InitializeCore(testConfig.DynamoCorePath);
```
Comment (Benglin) [inline test/Libraries/TestServices/AssemblyResolver.cs]: I'll have some merging to do, but basically this is what going to happen:

```csharp
// assemblyHelper is a data member of `AssemblyResolver` (which is 
// no longer static as of new Path Manager code base).
assemblyHelper = new AssemblyHelper(testConfig.DynamoCorePath);
AppDomain.CurrentDomain.AssemblyResolve += assemblyHelper.ResolveAssembly;
```

File: src/DynamoCore/Models/DynamoModel.cs [hunk 5]
```diff
@@ -545,7 +558,7 @@ private void OnAsyncTaskStateChanged(DynamoScheduler sender, TaskStateChangedEve
         /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
         /// </summary>
         /// <filterpriority>2</filterpriority>
-        public void Dispose()
+        public virtual void Dispose()
```
Comment (Benglin) [inline src/DynamoCore/Models/DynamoModel.cs]: I don't see why this needs to be made `virtual` (there's no changes for `RevitDynamoModel`). The derived class can completely bypass this by not calling `base.Dispose`, which is dangerous. If this really needs to be `virtual`, then I would recommend the following pattern:

```csharp
public void Dispose()
{
    DisposeCore();
    
    // Do the stuff we do today. The following clean-up codes are guaranteed 
    // to be executed even if derived class does not call base.DisposeCore().
    // 
    LibraryServices.Dispose();
    LibraryServices.LibraryManagementCore.__TempCoreHostForRefactoring.Cleanup();
    Logger.Dispose();
    
    // Bla bla bla...
}

protected virtual void DisposeCore()
{
    // Does nothing here...
}
```

**Comment by Benglin** [comment]
@ikeough, generally this looks nice and streamlined. I am not entirely certain what purposes does `IStartConfiguration` interface serve, the `DefaultStartConfiguration` is having all the default values (I can understand if it is used to encapsulate operations that repeatedly done by clients, but that doesn't seem to be the case).

Apart from that, LGTM.

**Comment by Benglin** [comment]
I think #242 makes the interface purposes clearer, LGTM.

File: src/DynamoCore/Models/DynamoModel.cs (hunk 5 above)
Comment (ikeough) [inline src/DynamoCore/Models/DynamoModel.cs]: There were changes to RevitDynamoModel, then they got backed out. Thanks for noticing this Ben.
